ci, gha: Run "x86_64: macOS Ventura" job on GitHub Actions

hebasto opened the pull request:
This PR solves one item in https://github.com/bitcoin-core/secp256k1/issues/1392.

hebasto commented:
> For macOS, we need to take also #1153 into account.

https://github.com/bitcoin-core/secp256k1/issues/1153#issue-1450259053:
> maybe it makes sense to reduce the number of MacOS jobs a little bit.

I'm opening for suggestion regarding that.

real-or-random commented:
The problem with this is that it removes all our tests on native ARM. 

We could instead run some Linux tasks on ARM, using Bitcoin Core's ARM runner on Cirrus. But if we do this, then this makes it less much attractive to move Linux x86_64 tasks to GitHub Actions. I don't know, it's all related to each other. :/

hebasto commented:
> The problem with this is that it removes all our tests on native ARM.

Want me to drop the second commit?

hebasto commented:
Converting to draft for now.

Please review https://github.com/bitcoin-core/secp256k1/pull/1389 first.

real-or-random commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,102 @@
+
+    strategy:
+      fail-fast: false
+      matrix:
+        include:
+          - { widemul:  'int64', recovery: 'yes', ecdh: 'yes', schnorrsig: 'yes', ellswift: 'yes' }
```
According to https://stackoverflow.com/a/75447432  you can use
```
      matrix:
        include:
          env: ...
```

AFAIU this will avoid all the `env` settings like `BUILD: ${{ matrix.build != '' && matrix.build || env.BUILD }}` below.

hebasto replied:
Reworked.

hebasto commented:
Related: https://github.com/github/roadmap/issues/528.

hebasto commented:
Updated 10037078145b2b94b8fbd49c1a7c7c0aa5ac31b3 -> 59ae09ba4838634b486fdf630e0befa0544941cc ([pr1394.02](https://github.com/hebasto/secp256k1/commits/pr1394.02) -> [pr1394.03](https://github.com/hebasto/secp256k1/commits/pr1394.03), [diff](https://github.com/hebasto/secp256k1/compare/pr1394.02..pr1394.03)):

- reworked this based on feedback from https://github.com/bitcoin-core/secp256k1/pull/1398

hebasto commented:
Updated 59ae09ba4838634b486fdf630e0befa0544941cc -> 10180bf7cb50ddcdfe0f6c50a85f0b63f42f0708 ([pr1394.03](https://github.com/hebasto/secp256k1/commits/pr1394.03) -> [pr1394.04](https://github.com/hebasto/secp256k1/commits/pr1394.04), [diff](https://github.com/hebasto/secp256k1/compare/pr1394.03..pr1394.04)):

- dropped the `Set environment variables` step in favor of direct `env: ${{ matrix.env_vars }}` assignment

hebasto commented:
> This uses the wrong GCC, namely Clang: [bitcoin-core/secp256k1/actions/runs/5841706589/job/15842131976?pr=1394#step:4:43](https://github.com/bitcoin-core/secp256k1/actions/runs/5841706589/job/15842131976?pr=1394#step:4:43)
> 
> But we have the same bug on Cirrus CI: 😞 [cirrus-ci.com/task/5276055561306112?logs=test#L17](https://cirrus-ci.com/task/5276055561306112?logs=test#L17)
> 
> This is so easy to get wrong on macOS. I suggest we just replace the gcc symlink to point to the actual gcc...

Fixed here.

Also see https://github.com/bitcoin-core/secp256k1/pull/1402.

real-or-random commented:
I'm currently looking into the failure, trying to figure out the right path. :)

hebasto commented:
> I'm currently looking into the failure, trying to figure out the right path. :)

Doing the same on my macOS :)

hebasto commented on .github/workflows/ci.yml:
```diff
@@ -8,9 +8,94 @@ on:
+          HOMEBREW_NO_INSTALL_CLEANUP: 1
+        run: |
+          brew install automake libtool gcc
+          ln -s /usr/local/bin/gcc-?? /usr/local/bin/gcc
```
Fixed.

real-or-random replied:
Oh sure, they even say it in the docs...

We could use a (simplified) variant of this: https://github.com/bitcoin-core/secp256k1/blob/b327abfcea90394b5c63890406d6a5c54d02212e/ci/linux-debian.Dockerfile#L52-L57 

That is, this should do it:
```
ln -s $(ls /usr/local/bin/gcc-?? | sort | tail -1) /usr/bin/local/bin/gcc)
```

real-or-random reviewed (approved):
ACK 8e54a346d2fa5aeedd6ba5201fcb084c281cf6a7 

That's really clean now, I think.

real-or-random commented:
@ Reviewers:

I think what still needs conceptual discussion when it comes to macOS is whether 
 - we follow the approach [I suggested above](https://github.com/bitcoin-core/secp256k1/pull/1394#issuecomment-1671784065), namely "moving" some of the existing macOS jobs to self-hosted ARM Linux Cirrus, in order to keep some ARM testing,
 - or whether we do something else, e.g., add just ARM QEMU jobs. [It's possible to run both the ctimetests (using MSan) and the normal tests in QEMU](https://github.com/bitcoin-core/secp256k1/issues/1392#issuecomment-1679306773). The only caveat is that we can run the ctimetests only for gcc, because gcc has no MSan.

Not entirely sure. Having everything on GHA., and not depending on the self-hosted runners, makes our infrastructure simpler. But on the other hand, having the cttimetest on GCC/ARM is probably a good idea. 

Anyway, all of this should probably happen after this PR here.

jonasnick reviewed (commented):
ACK 8e54a346d2fa5aeedd6ba5201fcb084c281cf6a7

real-or-random commented:
> I vote for keeping this single job on Cirrus for now

@jonasnick pointed out to me that we don't run the cttimetests on ARM currently (not before and not after this PR). I lost track of that apparently when I wrote that above...  (Sorry, to many platforms and special cases...)  To clarify, this means that removing the existing macOS tasks on Cirrus would be a loss only in the sense that run fewer *normal* tests on ARM. What we currently do for ARM is to run some tests in QEMU. My suggestion is to run more tasks in QEMU to mitigate this loss. We currently run them on Cirrus, but GHA should work equally.

Independently of this, it still makes sense to add a job that runs the cttimetests on ARM, and the only feasible way to do this right now is using Cirrus CI. I'll pick up the work #970 to implement this.

real-or-random commented:
> So I guess the Cirrus macos task can be removed, and a comment be added to switch to M1, once it is there?

Yes, indeed. See #1404.
